File: Leetcode/Python/_85.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        def calc_area(row):
            n = len(row)
            left = [0] * n
            right = [0] * n
            stack = []
            for i in range(n):
                if not stack:
                    stack.append(i)
                    left[i] = i
                else:
                    while stack and row[stack[-1]] >= row[i]:
                        stack.pop()
                    if stack:
                        left[i] = stack[-1] + 1
                    else:
                        left[i] = 0
                    stack.append(i)
            stack = []
            for i in range(n - 1, -1, -1):
                if not stack:
                    stack.append(i)
                    right[i] = i
                else:
                    while stack and row[stack[-1]] >= row[i]:
                        stack.pop()
                    if stack:
                        right[i] = stack[-1] - 1
                    else:
                        right[i] = n - 1
                    stack.append(i)
            area = 0
            for i in range(n):
                area = max(area, (right[i] - left[i] + 1) * row[i])
            return area
        output = [0] * len(matrix[0])
        area = 0
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                num = int(matrix[i][j])
                if num != 0:
                    output[j] = output[j] + int(matrix[i][j])
                else:
                    output[j] = 0
            temp = calc_area(output)
            area = max(area, temp)
        return area
                

File: Leetcode/cpp/85.cpp
#include <iostream>
#include <algorithm>
#include <vector>


class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int rows = matrix.size();
        if (rows == 0) return 0;
        int cols = matrix[0].size();
        int max_area = 0;
        std::vector<std::vector<int>> dp(rows, std::vector<int>(cols, 0));
        for (int row = 0; row < rows; ++row) {
            for (int col = 0; col < cols; ++col) {
                if (matrix[row][col] == '1') {
                    dp[row][col] = col == 0 ? 1 : dp[row][col - 1] + 1;
                    int width = dp[row][col];
                    for (int k = row; k >= 0; --k) {
                        width = std::min(width, dp[k][col]);
                        max_area = std::max(max_area, width * (row - k + 1));
                    }
                }
            }
        }
        return max_area;
    }
};
